File: BleModule/src/main/java/com/topdon/ble/Connection.java
package com.topdon.ble;

import android.bluetooth.*;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.UUID;

public interface Connection {
    UUID clientCharacteristicConfig = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb");

    int REQUEST_FAIL_TYPE_REQUEST_FAILED = 0;
    int REQUEST_FAIL_TYPE_CHARACTERISTIC_NOT_EXIST = 1;
    int REQUEST_FAIL_TYPE_DESCRIPTOR_NOT_EXIST = 2;
    int REQUEST_FAIL_TYPE_SERVICE_NOT_EXIST = 3;

    int REQUEST_FAIL_TYPE_GATT_STATUS_FAILED = 4;
    int REQUEST_FAIL_TYPE_GATT_IS_NULL = 5;
    int REQUEST_FAIL_TYPE_BLUETOOTH_ADAPTER_DISABLED = 6;
    int REQUEST_FAIL_TYPE_REQUEST_TIMEOUT = 7;
    int REQUEST_FAIL_TYPE_CONNECTION_DISCONNECTED = 8;
    int REQUEST_FAIL_TYPE_CONNECTION_RELEASED = 9;

    //-------------------
    int TIMEOUT_TYPE_CANNOT_DISCOVER_DEVICE = 0;

    int TIMEOUT_TYPE_CANNOT_CONNECT = 1;

    int TIMEOUT_TYPE_CANNOT_DISCOVER_SERVICES = 2;

    //--------------------------------

    int CONNECT_FAIL_TYPE_MAXIMUM_RECONNECTION = 1;

    int CONNECT_FAIL_TYPE_CONNECTION_IS_UNSUPPORTED = 2;

    @NonNull
    Device getDevice();

    int getMtu();

    void reconnect();

    void disconnect();

    void refresh();

    void release();

    void releaseNoEvent();

    @NonNull
    ConnectionState getConnectionState();

    boolean isAutoReconnectEnabled();

    @Nullable
    BluetoothGatt getGatt();

    void clearRequestQueue();

    void clearRequestQueueByType(RequestType type);

    @NonNull
    ConnectionConfiguration getConnectionConfiguration();

    @Nullable
    BluetoothGattService getService(UUID service);

    @Nullable
    BluetoothGattCharacteristic getCharacteristic(UUID service, UUID characteristic);

    @Nullable
    BluetoothGattDescriptor getDescriptor(UUID service, UUID characteristic, UUID descriptor);

    void execute(Request request);

    boolean isNotificationOrIndicationEnabled(BluetoothGattCharacteristic characteristic);

    boolean isNotificationOrIndicationEnabled(UUID service, UUID characteristic);

    void setBluetoothGattCallback(BluetoothGattCallback callback);

    boolean hasProperty(UUID service, UUID characteristic, int property);
}

File: BleModule/src/main/java/com/mpdc4gsr/ble/Connection.kt
package com.mpdc4gsr.ble

import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothGattService
import java.util.UUID

interface Connection {
    val device: Device

    val mtu: Int

    fun reconnect()

    fun disconnect()

    fun refresh()

    fun release()

    fun releaseNoEvent()

    val connectionState: ConnectionState

    val isAutoReconnectEnabled: Boolean

    val gatt: BluetoothGatt?

    fun clearRequestQueue()

    fun clearRequestQueueByType(type: RequestType?)

    val connectionConfiguration: ConnectionConfiguration

    fun getService(service: UUID?): BluetoothGattService?

    fun getCharacteristic(service: UUID?, characteristic: UUID?): BluetoothGattCharacteristic?

    fun getDescriptor(service: UUID?, characteristic: UUID?, descriptor: UUID?): BluetoothGattDescriptor?

    fun execute(request: Request?)

    fun isNotificationOrIndicationEnabled(characteristic: BluetoothGattCharacteristic?): Boolean

    fun isNotificationOrIndicationEnabled(service: UUID?, characteristic: UUID?): Boolean

    fun setBluetoothGattCallback(callback: BluetoothGattCallback?)

    fun hasProperty(service: UUID?, characteristic: UUID?, property: Int): Boolean

    companion object {
        val clientCharacteristicConfig: UUID? = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")

        const val REQUEST_FAIL_TYPE_REQUEST_FAILED: Int = 0
        const val REQUEST_FAIL_TYPE_CHARACTERISTIC_NOT_EXIST: Int = 1
        const val REQUEST_FAIL_TYPE_DESCRIPTOR_NOT_EXIST: Int = 2
        const val REQUEST_FAIL_TYPE_SERVICE_NOT_EXIST: Int = 3

        const val REQUEST_FAIL_TYPE_GATT_STATUS_FAILED: Int = 4
        const val REQUEST_FAIL_TYPE_GATT_IS_NULL: Int = 5
        const val REQUEST_FAIL_TYPE_BLUETOOTH_ADAPTER_DISABLED: Int = 6
        const val REQUEST_FAIL_TYPE_REQUEST_TIMEOUT: Int = 7
        const val REQUEST_FAIL_TYPE_CONNECTION_DISCONNECTED: Int = 8
        const val REQUEST_FAIL_TYPE_CONNECTION_RELEASED: Int = 9
        const val REQUEST_FAIL_TYPE_NO_PERMISSION: Int = 10

        const val TIMEOUT_TYPE_CANNOT_DISCOVER_DEVICE: Int = 0

        const val TIMEOUT_TYPE_CANNOT_CONNECT: Int = 1

        const val TIMEOUT_TYPE_CANNOT_DISCOVER_SERVICES: Int = 2


        const val CONNECT_FAIL_TYPE_MAXIMUM_RECONNECTION: Int = 1

        const val CONNECT_FAIL_TYPE_CONNECTION_IS_UNSUPPORTED: Int = 2

        const val CONNECT_FAIL_TYPE_NO_PERMISSION: Int = 3
    }
}
